From 3db978d582bc056fa915a91c65af279bb406924c Mon Sep 17 00:00:00 2001
From: Olliver Schinagl <oliver@schinagl.nl>
Date: Wed, 14 Aug 2024 09:40:17 +0200
Subject: [PATCH 18/18] sdk: Fake IEE Address response for Xiaomi devices

It appears that Xiaomi devices won't work with coordinators that are not
of Xiaomi. To counter this, pretend we're a Xiaomi device by faking the
IEE Address.

See https://github.com/Koenkk/zigbee2mqtt/issues/9274 for details.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>
---
 source/ti/zstack/stack/zdo/zd_object.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/source/ti/zstack/stack/zdo/zd_object.c b/source/ti/zstack/stack/zdo/zd_object.c
index 0c62d58d6..d1f899cec 100644
--- a/source/ti/zstack/stack/zdo/zd_object.c
+++ b/source/ti/zstack/stack/zdo/zd_object.c
@@ -676,6 +676,20 @@ void ZDO_ProcessNodeDescReq( zdoIncomingMsg_t *inMsg )
 
   if ( desc != NULL )
   {
+    uint8_t extAddr[Z_EXTADDR_LEN];
+    // Respond with Xiaomi manufacturer code when ieeAddr is withing Xiaomi address space
+    // Otherwise some devices don't work
+    // https://github.com/Koenkk/zigbee2mqtt/issues/9274
+    if (APSME_LookupExtAddr(inMsg->srcAddr.addr.shortAddr, extAddr) == TRUE &&
+         ((extAddr[7] == 0x04 && extAddr[6] == 0xcf && extAddr[5] == 0x8c) ||
+          (extAddr[7] == 0x54 && extAddr[6] == 0xef && extAddr[5] == 0x44))) {
+        desc->ManufacturerCode[0] = 0x5f;
+        desc->ManufacturerCode[1] = 0x11;
+    } else {
+        desc->ManufacturerCode[0] = 0x0;
+        desc->ManufacturerCode[1] = 0x0;
+    }
+
     ZDP_NodeDescMsg( inMsg, aoi, desc );
   }
   else
-- 
2.46.0

